pkg/storage/stores/shipper/downloads/table.go

```diff
@@ -0,0 +1,400 @@
+	table := Table{
+		name:              name,
+		cacheLocation:     cacheLocation,
+		metrics:           metrics,
+		storageClient:     storageClient,
+		boltDBIndexClient: boltDBIndexClient,
+		dbs:               map[string]*downloadedFile{},
+		ready:             make(chan struct{}),
+	}
+
+	// keep the files collection locked until all the files are downloaded.
+	table.dbsMtx.Lock()
+	go func() {
+		defer table.dbsMtx.Unlock()
+		defer close(table.ready)
+
+		ctx, cancel := context.WithTimeout(context.Background(), downloadTimeout)
+		defer cancel()
+
+		table.cancelFunc = cancel
+
+		// Using background context to avoid cancellation of download when request times out.
+		// We would anyways need the files for serving next requests.
+		if err := table.init(ctx); err != nil {
+			level.Error(util.Logger).Log("msg", "failed to download files", "name", table.name)
+		}
+	}()
+
+	return &table
+}
+
+// init downloads all the db files for the table from object storage.
+// it assumes the locking of mutex is taken care of by the caller.
+func (t *Table) init(ctx context.Context) (err error) {
+	defer func() {
+		status := statusSuccess
+		if err != nil {
+			status = statusFailure
+			t.err = err
+
+			// cleaning up files due to error to avoid returning invalid results.
+			for fileName := range t.dbs {
+				if err := t.cleanupDB(fileName); err != nil {
+					level.Error(util.Logger).Log("msg", "failed to cleanup partially downloaded file", "filename", fileName, "err", err)
+				}
+			}
+		}
+		t.metrics.filesDownloadOperationTotal.WithLabelValues(status).Inc()
+	}()
+
+	startTime := time.Now()
+	totalFilesSize := int64(0)
+
+	objects, _, err := t.storageClient.List(ctx, t.name+"/")
+	if err != nil {
+		return
+	}
+
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf("list of files to download for period %s: %s", t.name, objects))
+
+	folderPath, err := t.folderPathForTable(true)
+	if err != nil {
+		return
+	}
+
+	for _, object := range objects {
+		dbName, err := getDBNameFromObjectKey(object.Key)
+		if err != nil {
+			return err
+		}
+
+		filePath := path.Join(folderPath, dbName)
+		df := downloadedFile{}
+
+		err = t.getFileFromStorage(ctx, object.Key, filePath)
+		if err != nil {
+			return err
+		}
+
+		df.mtime = object.ModifiedAt
+		df.boltdb, err = local.OpenBoltdbFile(filePath)
+		if err != nil {
+			return err
+		}
+
+		var stat os.FileInfo
+		stat, err = os.Stat(filePath)
+		if err != nil {
+			return err
+		}
+
+		totalFilesSize += stat.Size()
+
+		t.dbs[dbName] = &df
+	}
+
+	duration := time.Since(startTime).Seconds()
+	t.metrics.filesDownloadDurationSeconds.add(t.name, duration)
+	t.metrics.filesDownloadSizeBytes.add(t.name, totalFilesSize)
+
+	return
+}
+
+// Closes references to all the dbs.
+func (t *Table) Close() {
+	t.dbsMtx.Lock()
+	defer t.dbsMtx.Unlock()
+
+	// stop the initialization if it is still ongoing.
+	t.cancelFunc()
```

Comment:
The problem here is the lock will prevent you from actually cancelling.

I think you should use a different cancellation than the timeout, one that is initialized not the init goroutine, this way you don't race to cancel (that's why you need the lock) the first cancellation will also trigger the timeout context.

Reply:
See patch example:


```diff
diff --git a/pkg/storage/stores/shipper/downloads/table.go b/pkg/storage/stores/shipper/downloads/table.go
index 2107a589..99ee8603 100644
--- a/pkg/storage/stores/shipper/downloads/table.go
+++ b/pkg/storage/stores/shipper/downloads/table.go
@@ -49,6 +49,7 @@ type Table struct {
 }

 func NewTable(name, cacheLocation string, storageClient chunk.ObjectClient, boltDBIndexClient BoltDBIndexClient, metrics *metrics) *Table {
+	ctx, cancel := context.WithCancel(context.Background())
 	table := Table{
 		name:              name,
 		cacheLocation:     cacheLocation,
@@ -57,6 +58,7 @@ func NewTable(name, cacheLocation string, storageClient chunk.ObjectClient, bolt
 		boltDBIndexClient: boltDBIndexClient,
 		dbs:               map[string]*downloadedFile{},
 		ready:             make(chan struct{}),
+		cancelFunc:        cancel,
 	}

 	// keep the files collection locked until all the files are downloaded.
@@ -65,11 +67,9 @@ func NewTable(name, cacheLocation string, storageClient chunk.ObjectClient, bolt
 		defer table.dbsMtx.Unlock()
 		defer close(table.ready)

-		ctx, cancel := context.WithTimeout(context.Background(), downloadTimeout)
+		ctx, cancel := context.WithTimeout(ctx, downloadTimeout)
 		defer cancel()

-		table.cancelFunc = cancel
-
 		// Using background context to avoid cancellation of download when request times out.
 		// We would anyways need the files for serving next requests.
 		if err := table.init(ctx); err != nil {
@@ -154,12 +154,11 @@ func (t *Table) init(ctx context.Context) (err error) {

 // Closes references to all the dbs.
 func (t *Table) Close() {
-	t.dbsMtx.Lock()
-	defer t.dbsMtx.Unlock()
-
 	// stop the initialization if it is still ongoing.
 	t.cancelFunc()

+	t.dbsMtx.Lock()
+	defer t.dbsMtx.Unlock()
 	for name, db := range t.dbs {
 		if err := db.boltdb.Close(); err != nil {
 			level.Error(util.Logger).Log("msg", fmt.Errorf("failed to close file %s for table %s", name, t.name))

```